File: 332. Reconstruct Itinerary.py
# 332. Reconstruct Itinerary

class Solution:
    def findItinerary(self, tickets: List[List[str]]) -> List[str]:
        adj = defaultdict(list)
        for src, dest in sorted(tickets, reverse = True):
            adj[src].append(dest)
        
        ans = []
        def dfs(node):
            while adj[node]:
                next_node = adj[node].pop()
                dfs(next_node)
            ans.append(node)
        
        dfs('JFK')
        return ans[::-1]

        # graph = defaultdict(list)

        # # Build the graph as a priority queue (min-heap) for lexical order
        # for src, dest in sorted(tickets):
        #     graph[src].append(dest)

        # result = []

        # def dfs(node):
        #     while graph[node]:
        #         next_dest = graph[node].pop(0)
        #         dfs(next_dest)
        #     result.append(node)

        # # Start the journey from 'JFK'
        # dfs('JFK')

        # # Since the result is constructed in reverse order, reverse it
        # return result[::-1]

File: 332. Reconstruct Itinerary.cpp
// 332. Reconstruct Itinerary
#include <vector>
#include <string>
#include <unordered_map>
#include <algorithm>

using namespace std;

class Solution {
    public:
        vector<string> findItinerary(vector<vector<string>>& tickets) {
            unordered_map<string, vector<string>> map;
    
            for (auto& t : tickets) {
                map[t[0]].push_back(t[1]);
            }
    
            for (auto& [src, des] : map) {
                sort(des.rbegin(), des.rend());
            }
    
            vector<string> ans;
            travel("JFK", map, ans);
            reverse(ans.begin(), ans.end());
            return ans;
        }
    
    private:
        void travel(string node, unordered_map<string, vector<string>>& map, vector<string>& ans) {
            while (!map[node].empty()) {
                string next_node = map[node].back();
                map[node].pop_back();
                travel(next_node, map, ans);
            }
            ans.push_back(node);
        }
    };

